### PR TITLE
Add service based gcs client

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
Service based gcs client connect to gcs service and support gcs operation.
It will replace redis gcs client.
This pr change c++ and java testcase, python testcase will be changed in next pr.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_server.h]: document this method

File: src/ray/raylet/raylet.cc
Comment (raulchen) [inline src/ray/raylet/raylet.cc]: Consolidate this with the gcs client initialization code in `main.cc`.

Comment (ffbin) [reply src/ray/raylet/raylet.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.h
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.h]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.h]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: AddJob() is void

File: src/ray/gcs/gcs_client/service_based_accessor.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.h
Comment (ashione) [inline src/ray/gcs/gcs_client/service_based_accessor.h]: Why not move this mutable initialization to constructor list.

File: java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java]: This should be put in the above `if isHead` section. Also, can you consolidate this with `startRedisServer` into a method called `startGcs`?

File: java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java]: Add `gcs_service_enabled` to `rayletConfigParameters` before `String.join`

**Review by raulchen — CHANGES_REQUESTED** [review]
talked with @ffbin offline. We should add a env var to control using gcs server or not, and enable it in a new CI job that runs integration tests. So we don't change any default behavior now. When GCS server is completely ready, we'll switch the default option to GCS sever, and remove redis-based client as well as the new CI job.

File: src/ray/common/ray_config_def.h
Comment (zhijunfu) [inline src/ray/common/ray_config_def.h]: move this above the two configurations for gcs service

File: src/ray/common/ray_config_def.h
Comment (zhijunfu) [inline src/ray/common/ray_config_def.h]: add a comment that this only takes effect when gcs service is enabled.

File: src/ray/core_worker/test/core_worker_test.cc
Comment (zhijunfu) [inline src/ray/core_worker/test/core_worker_test.cc]: sounds like gcs server is not used in this test?

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (zhijunfu) [inline src/ray/gcs/gcs_server/gcs_server.cc]: remove this

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (zhijunfu) [inline src/ray/gcs/gcs_server/gcs_server.h]: would you add some details on why this is required? e.g. to let clients to get new gcs server ip address if it crashes ...

File: src/ray/raylet/main.cc
Comment (zhijunfu) [inline src/ray/raylet/main.cc]: DEBUG?

File: src/ray/test/run_core_worker_tests.sh
Comment (zhijunfu) [inline src/ray/test/run_core_worker_tests.sh]: why is this needed?

**Review by zhijunfu — COMMENTED** [review]
Agreed with @raulchen then we should add a new CI for service based client and keep gcs service disabled for now. The gcs service unit test itself is not sufficient to guarantee its correctness.  I also left a few minor comments. Thanks.

File: src/ray/raylet/raylet.cc
Comment (ffbin) [reply src/ray/raylet/raylet.cc]: fixed

File: java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java
Comment (ffbin) [reply java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.h
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.h]: fixed

File: java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java
Comment (ffbin) [reply java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java]: fixed

File: src/ray/common/ray_config_def.h
Comment (ffbin) [reply src/ray/common/ray_config_def.h]: use a env var instead of config now.

File: src/ray/common/ray_config_def.h
Comment (ffbin) [reply src/ray/common/ray_config_def.h]: fixed

File: src/ray/test/run_core_worker_tests.sh
Comment (ffbin) [reply src/ray/test/run_core_worker_tests.sh]: if we enable gcs service, we need start gcs server before start raylet.

File: src/ray/raylet/main.cc
Comment (ffbin) [reply src/ray/raylet/main.cc]: this log is removed because we use a env var instead of config now.

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.h]: fixed

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.cc]: fixed

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.h]: fixed

File: .travis.yml
Comment (raulchen) [inline .travis.yml]:
```suggestion
        - RAY_GCS_SERVICE_ENABLED=true
```

File: src/ray/core_worker/test/core_worker_test.cc
Comment (raulchen) [inline src/ray/core_worker/test/core_worker_test.cc]: Only start GCS Server when enabled

File: src/ray/gcs/gcs_client/service_based_accessor.cc
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.cc]: just one line `rpc::ActorTableData actor_table_data(reply.actor_table_data());`

Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/core_worker/test/core_worker_test.cc
Comment (ffbin) [reply src/ray/core_worker/test/core_worker_test.cc]: fixed

File: .travis.yml
Comment (ffbin) [reply .travis.yml]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

File: src/ray/raylet/main.cc
Comment (ffbin) [reply src/ray/raylet/main.cc]: fixed

File: streaming/src/test/run_streaming_queue_test.sh
Comment (ffbin) [reply streaming/src/test/run_streaming_queue_test.sh]: fixed

File: src/ray/core_worker/test/core_worker_test.cc
Comment (ffbin) [reply src/ray/core_worker/test/core_worker_test.cc]: fixed

File: src/ray/core_worker/test/core_worker_test.cc
Comment (ffbin) [reply src/ray/core_worker/test/core_worker_test.cc]: fixed

File: src/ray/core_worker/test/core_worker_test.cc
Comment (ffbin) [reply src/ray/core_worker/test/core_worker_test.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.h]: fixed

**Review by raulchen — APPROVED** [review]

**Review by zhijunfu — APPROVED** [review]
LGTM

**Comment by mehrdadn** [comment]
@raulchen @ffbin I think this pull request broke some things in the Windows build. Note that the Windows builds aren't on Travis, so that might be why you didn't notice them. They're in GitHub Actions, right above the Merge button. I've fixed the issues #6811, so no need to worry for this one, but in the future please check that the Windows builds are also passing before merging pull requests. Thanks!

**Comment by raulchen** [comment]
@mehrdadn sorry about this! I've just notice this issue and told @ffbin to fix this. 
Thanks for your fix. We'll be more careful in the future.

**Pull request merged**
